[FLINK-38822] Introduce URL_DECODE_RECURSIVE function for recursive URL decoding

featzhang opened the pull request:
## Motivation

The current `URL_DECODE` function only performs a single-pass URL decoding. In practice, URLs
can be double- or multi-encoded (e.g., `https%253A%252F%252F...`), requiring multiple rounds
of decoding to recover the original string.

## Changes

Introduce a new built-in scalar function `URL_DECODE_RECURSIVE` that recursively decodes a
URL-encoded string until the result stabilizes or a maximum decode depth is reached.

### New SQL Function
- `URL_DECODE_RECURSIVE(string, maxDepth)` - Recursively decode a URL-encoded string
- `string`: The URL-encoded string to decode (can be NULL)
- `maxDepth`: Optional integer specifying maximum decode iterations (default: 10, must be > 0)
  - Must be a literal integer (validated at compile time)
  - Non-literal values will cause a compilation error

### New Table API Method
- `STRING.urlDecodeRecursive()` - Uses default maxDepth=10
- `STRING.urlDecodeRecursive(maxDepth)` - Uses specified maxDepth

### Behavior
- Decoding stops when the output equals the input (stable) or maxDepth iterations are reached
- **If at least one successful decoding has been performed, returns the last successfully decoded result**
- Returns `NULL` if the input is `NULL`
- Returns the input value if no successful decoding occurs (e.g., first iteration fails)

### Implementation Notes
- The maxDepth parameter is validated at compile time via `MaxDepthArgumentTypeStrategy`
- This prevents runtime errors and ensures performance
- Default maxDepth of 10 prevents infinite loops on circular encodings
- Defensive checks removed from eval() since validation happens at planning time

## Examples
```sql
-- Simple decode with default maxDepth
SELECT URL_DECODE_RECURSIVE('Hello%20World');  -- 'Hello World'

-- Double-encoded URL with default maxDepth
SELECT URL_DECODE_RECURSIVE('Hello%2520World');  -- 'Hello World'

-- Triple-encoded URL with maxDepth=2 (stops after 2 decodes)
SELECT URL_DECODE_RECURSIVE('Hello%252520World', 2);  -- 'Hello%20World'

-- Triple-encoded URL with maxDepth=3 (completes all decodes)
SELECT URL_DECODE_RECURSIVE('Hello%252520World', 3);  -- 'Hello World'
```

## Notes

- Non-recursive usage is still covered by the existing `URL_DECODE` function
- The maxDepth parameter must be a literal integer, not a dynamic column value
- Compile-time validation provides early error detection and better performance
- This function is particularly useful for handling multi-encoded URLs in web crawling and data extraction scenarios

Fixed: FLINK-38822

lincoln-lil commented on docs/data/sql_functions.yml:
```diff
@@ -492,6 +492,13 @@ string:
     description:
       Decodes a given string in 'application/x-www-form-urlencoded' format using the UTF-8 encoding scheme.
       If the input is NULL, or there is an issue with the decoding process(such as encountering an illegal escape pattern), or the encoding scheme is not supported, the function returns NULL.
+  - sql: URL_DECODE_RECURSIVE(string, boolean)
```
If we decide to introduce a new function, the `boolean` param seems unnecessary because the function name already indicate the `recursive` behavior, and for the non-recursive usage users can use `url_decode` function, WDYT?

lincoln-lil reviewed (commented):
@featzhang Thanks for contributing this! I've left some comments there. Since we have effectively introduced a new function name, I suggest treating it as an entirely new function in terms of naming, including the JIRA title.

dylanhz commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/UrlDecodeFunction.java:
```diff
@@ -30,22 +31,78 @@
-            return StringData.fromString(URLDecoder.decode(value.toString(), charset.name()));
+            while (iteration < maxDepth) {
+                previousValue = currentValue;
+                currentValue = URLDecoder.decode(currentValue, charset.name());
```
nit：We can use `URLDecoder.decode(String, Charset)` since JDK 8 is no longer supported, and remove `UnsupportedEncodingException` catching as well.

dylanhz reviewed (changes requested):
@featzhang Thanks for contributing this! I've left some comments there. Also, please add corresponding tests for the new methods in `test_expression.py`.

featzhang commented:
@flinkbot run azure

featzhang commented:
@flinkbot run azure

dylanhz reviewed (commented):
@featzhang Thanks for the feedback. I have only two questions:
1. The doc says "the function returns NULL" on decoding errors, but the implementation returns the last successful intermediate result when `iteration > 0`.
2. Since `maxDepth` is already constrained as `LITERAL` in the input type strategy, its value is known at planning time. The validation should be moved to type inference — reject non-positive values with a clear compile-time error, rather than silently falling back to the default of `10` at runtime. This way the eval method doesn't need any defensive checks on `maxDepth`.

dylanhz reviewed (commented):
@featzhang Thanks for addressing the previous feedback! Left a few comments. Also, don't forget to update the doc as mentioned before.

featzhang commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/MaxDepthArgumentTypeStrategy.java:
```diff
@@ -0,0 +1,86 @@
+            CallContext callContext, int argumentPos, boolean throwOnFailure) {
+        final var actualType = callContext.getArgumentDataTypes().get(argumentPos).getLogicalType();
+
+        if (!actualType.is(LogicalTypeFamily.INTEGER_NUMERIC)) {
```
Support for different integer types: I've updated MaxDepthArgumentTypeStrategy to accept all integer types (TINYINT, SMALLINT, INT, BIGINT) instead of just INT. The implementation now uses DataTypes.INTEGRAL() to match any integral type.

featzhang committed:
feat: [FLINK-38822][Table] Support URL_DECODE and URL_DECODE_RECURSIVE functions

- Add URL_DECODE and URL_DECODE_RECURSIVE built-in functions
- Support multiple integer types (TINYINT, SMALLINT, INT, BIGINT) for maxDepth parameter
- Add compile-time validation for maxDepth parameter via MaxDepthArgumentTypeStrategy
- Expose maxDepth() factory method in SpecificInputTypeStrategies to reduce API surface
- Add ITCase tests for URL_DECODE_RECURSIVE with different integer types
- Add Python tests and improve documentation for URL_DECODE_RECURSIVE
- Fix PyFlink test: use explicit DataTypes schema to handle None column in from_elements

featzhang committed:
[FLINK-38822] Add connection pool management options to TritonOptions and regenerate docs

- Add connection pool configuration options:
  - connection-pool-max-idle: Maximum idle connections (default: 20)
  - connection-pool-keep-alive: Keep-alive duration (default: 5 min)
  - connection-pool-max-total: Max total connections (default: 100)
  - connection-timeout: Connection timeout (default: 10s)
  - connection-reuse-enabled: Enable connection reuse (default: true)
  - connection-pool-monitoring-enabled: Enable pool monitoring (default: false)
- Regenerate configuration documentation

featzhang committed:
[FLINK-38822] Add missing import for MaxDepthArgumentTypeStrategy in InputTypeStrategiesTest

featzhang committed:
[FLINK-38822] Fix code formatting in UrlFunctionsITCase

featzhang committed:
[FLINK-38822] Fix UrlFunctionsITCase tests: use testTableApiResult for TINYINT/SMALLINT/BIGINT literal tests

SQL CAST expressions (e.g., CAST(2 AS TINYINT)) are not considered
literals by the MaxDepthArgumentTypeStrategy, which requires compile-time
literal arguments. Changed the TINYINT/SMALLINT/BIGINT maxDepth tests to
use testTableApiResult instead of testResult to only test Table API,
since SQL does not support typed integer literals.

lincoln-lil commented on docs/layouts/shortcodes/generated/model_triton_advanced_section.html:
```diff
@@ -20,6 +20,42 @@
             <td>Compression algorithm for request body. Currently only <code class="highlighter-rouge">gzip</code> is supported. When enabled, the request body will be compressed to reduce network bandwidth.</td>
         </tr>
+        <tr>
+            <td><h5>connection-pool-keep-alive</h5></td>
```
These irrelevant changes shouldn't appear.

lincoln-lil commented on flink-models/flink-model-triton/src/main/java/org/apache/flink/model/triton/TritonOptions.java:
```diff
@@ -181,4 +181,95 @@ private TritonOptions() {
                                             code("'X-Custom-Header:value,X-Another:value2'"))
                                     .build());
+
+    // ========== Connection Pool Management Options ==========
```
ditto here

featzhang committed:
[FLINK-38822][Table] Address review feedback: remove unrelated changes, simplify MaxDepthArgumentTypeStrategy

- Remove unrelated Triton connector changes (model_triton_advanced_section.html,
  triton_configuration.html, TritonOptions.java)
- Simplify MaxDepthArgumentTypeStrategy using Number.class pattern
  (similar to PercentageArgumentTypeStrategy) while preserving original input types
- Restore original urlDecode/urlEncode Javadoc in BaseExpressions.java
- Add explicit Javadoc for recursive URL decode methods

lincoln-lil commented:
@featzhang Just a reminder, there’s an issue in Jira that needs to be followed up.
